src/composeappmanager.cc

```diff
@@ -475,23 +475,36 @@ data::InstallationResult ComposeAppManager::finalizeInstall(const Uptane::Target
-            app_pair.first % app_pair.second % run_res.err)};
-
-        LOG_ERROR << err_desc;
-        // Do we need to set some flag for the uboot and trigger a system reboot in order to boot on a previous
-        // ostree version, hence a proper/full rollback happens???
-        ir.description += ", however " + err_desc;
-        ir.description += "\n# Apps running:\n" + getRunningAppsInfoForReport();
-        // this is a hack to distinguish between ostree install (rollback) and App start failures.
-        // data::ResultCode::Numeric::kInstallFailed - boot on a new ostree version failed (rollback at boot)
-        // data::ResultCode::Numeric::kCustomError - boot on a new version was successful but new App failed to start
-        return data::InstallationResult(run_res.imagePullFailure() ? data::ResultCode::Numeric::kDownloadFailed
-                                                                   : data::ResultCode::Numeric::kCustomError,
-                                        ir.description);
+      if(!run_res) {
+        // First we print the run_res error message
+        LOG_ERROR << "App: " << app_pair.first << ", failed to start: " << run_res.err;
+        // If this fails it is possible that some apps are missing since they are added in this new target.
+        LOG_INFO << "Downloading App because it would not start";
+        Download(Target::toTufTarget(target));  // download the missing apps
+                                                // Retry the run
+        LOG_INFO << "Retrying to start App: " << app_pair.first;
+        const AppEngine::Result rerun_res = app_engine_->run({app_pair.first, app_pair.second});
+        if (!rerun_res) {
+        // again we print the run_res error message
+        LOG_ERROR << "App: " << app_pair.first << ", failed to start at rerun: " << rerun_res.err;
+          const std::string err_desc{boost::str(
+              boost::format("failed to start App after booting on a new sysroot version; app: %s; uri: %s; err: %s") %
+              app_pair.first % app_pair.second % run_res.err)};
+
+          LOG_ERROR << err_desc;
+          // Do we need to set some flag for the uboot and trigger a system reboot in order to boot on a previous
+          // ostree version, hence a proper/full rollback happens???
+          ir.description += ", however " + err_desc;
+          ir.description += "\n# Apps running:\n" + getRunningAppsInfoForReport();
+          // this is a hack to distinguish between ostree install (rollback) and App start failures.
+          // data::ResultCode::Numeric::kInstallFailed - boot on a new ostree version failed (rollback at boot)
+          // data::ResultCode::Numeric::kCustomError - boot on a new version was successful but new App failed to start
+          return data::InstallationResult(run_res.imagePullFailure() ? data::ResultCode::Numeric::kDownloadFailed
```

Comment:
Hm, this is problematic. Unfortunately, the code that invokes this method does not handle the case when `rerun_res.noSpace() == true`, it simply does not expect that the update "finalization" pulls any data. In this case we should NOT do rollback, while the current solution will do it if there is not enough free storage to accommodate new app...